Add modules and command to verify a tag against local content

ehelms opened the pull request:
Allows checking if what is local is in the tag on Koji and vice versa.

```
obal verify-tag --tag foreman-nightly-rhel7
obal verify-tag --all
```

@ekohl This would implement some of the gaps you requested

evgeni commented on obal/data/modules/verify_tag.py:
```diff
@@ -0,0 +1,65 @@
+        packages_in_koji = check_output(command, stderr=STDOUT, universal_newlines=True)
+        packages_in_koji = [item.split(' ')[0] for item in packages_in_koji.split("\n")]
+        packages_in_koji = [package for package in packages_in_koji if len(package) > 0]
+    except subprocess.CalledProcessError as error:
```
that's not imported and thus cannot be caught.

evgeni replied:
```suggestion
    except CalledProcessError as error:
```

evgeni commented on obal/data/modules/verify_tag.py:
```diff
@@ -0,0 +1,65 @@
+"""
+
+import os
+from subprocess import STDOUT, check_output
```
```suggestion
from subprocess import STDOUT, check_output, CalledProcessError
```

evgeni commented on obal/data/roles/verify_tag/tasks/main.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+- include_tasks: "verify.yaml"
+  when: verify_tag_all == False
```
`[E601] Don't compare to literal True/False` ;)

evgeni commented on obal/data/roles/verify_tag/tasks/main.yaml:
```diff
@@ -0,0 +1,6 @@
+  when: verify_tag_all == False
+
+- include_tasks: "all.yaml"
+  when: verify_tag_all == True
```
`[E601] Don't compare to literal True/False` ;)

evgeni commented:
Should the modules (not the command/playbook) be prefixed with `koji_` or so?

evgeni commented on tests/test_functional.py:
```diff
@@ -670,3 +670,10 @@ def test_obal_mock():
+def test_obal_verify_tag():
+    assert_obal_success(['verify-tag', '--tag', 'obaltest-nightly-rhel7'])
+
+    expected_log = []
```
that's not true ;)

evgeni commented on tests/fixtures/testrepo/upstream/package_manifest.yaml:
```diff
@@ -10,6 +10,8 @@ packages:
       - name: obaltest-nightly-rhel7
         dist: '.el7'
+      - name: obaltest-nightly-el8
+        dist: '.el8'
```
this made some other tests fail as they have not been updated to check for both el7 and el8 things:


```
FAILED tests/test_functional.py::test_obal_check_upstream_hello
FAILED tests/test_functional.py::test_obal_release_upstream_hello
FAILED tests/test_functional.py::test_obal_release_upstream_hello_nowait
FAILED tests/test_functional.py::test_obal_release_upstream_hello_nowaitrepo
```

evgeni commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,70 @@
+
+    try:
+        command = ['koji', 'list-pkgs', '--quiet', '--tag', tag_name]
+        packages_in_koji = subprocess.check_output(
```
any reason not to use `koji_wrapper.koji` that we have?

ekohl replied:
I generally don't like changing a variable in this way. First it's a string (after calling `check_output`) and then it becomes a list. That makes it hard to reason about it.
```suggestion
        koji_output = subprocess.check_output(
```

evgeni commented:
nitpicks and test-failures, but overall :+1:

ekohl commented on obal/data/modules/koji_find_tags.py:
```diff
@@ -0,0 +1,31 @@
+
+    packages = module.params['packages']
+
+    tags = []
```
I'd suggest to make this a `set()`
```suggestion
    tags = set()
```

ekohl commented on obal/data/modules/koji_find_tags.py:
```diff
@@ -0,0 +1,31 @@
+
+    for (_package, attributes) in packages.items():
+        if 'koji_tags' in attributes:
+            for tag in attributes['koji_tags']:
+                if tag['name'] not in tags:
+                    tags.append(tag['name'])
```
```suggestion
            tags += set(attributes['koji_tags'])
```

ekohl commented on obal/data/modules/koji_find_tags.py:
```diff
@@ -0,0 +1,31 @@
+
+    tags = []
+
+    for (_package, attributes) in packages.items():
```
```suggestion
    for attributes in packages.values():
```

ekohl commented on obal/data/modules/koji_find_tags.py:
```diff
@@ -0,0 +1,31 @@
+                if tag['name'] not in tags:
+                    tags.append(tag['name'])
+
+    module.exit_json(changed=False, tags=tags)
```
```suggestion
    module.exit_json(changed=False, tags=sorted(tags))
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,70 @@
+    except subprocess.CalledProcessError as error:
+        module.fail_json(changed=False, msg=error.output)
+
+    missing_in_koji = []
+    missing_in_git = []
+
+    for package in packages_for_tag:
+        if package not in packages_in_koji:
+            missing_in_koji.append(package)
+
+    for package in packages_in_koji:
+        if package not in packages_for_tag:
+            missing_in_git.append(package)
```
Sets are your friend
```suggestion
    missing_in_koji = packages_for_tag - packages_in_koji
    missing_in_git = packages_in_koji - packages_for_tag
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,70 @@
+
+    tag_name = module.params['tag']
+
+    packages_for_tag = []
```
```suggestion
    packages_for_tag = set()
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,70 @@
+                specfile = "%s/%s/%s.spec" % (attributes['package_base_dir'], package, package)
+                scl = tag['scl'] if 'scl' in tag.keys() else None
+                name = get_specfile_name(os.path.join(module.params['directory'], specfile), scl)
+                packages_for_tag.append(name)
```
```suggestion
                packages_for_tag.add(name)
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,70 @@
+        if package not in packages_for_tag:
+            missing_in_git.append(package)
+
+    module.exit_json(changed=False, missing_in_git=missing_in_git, missing_in_koji=missing_in_koji)
```
```suggestion
    module.exit_json(changed=False, missing_in_git=sorted(missing_in_git), missing_in_koji=sorted(missing_in_koji))
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,70 @@
+            universal_newlines=True
+        )
+
+        packages_in_koji = [item.split(' ')[0] for item in packages_in_koji.split("\n")]
+        packages_in_koji = [package for package in packages_in_koji if len(package) > 0]
```
In Python you don't need `len()` here since empty strings are also `False`. I'd also say that either a line is empty or there's something on it, so you can optimize here. Also, sets are great
```suggestion
        packages_in_koji = {item.split(' ')[0] for item in packages_in_koji.split("\n") if item}
```

This will be different if there are lines where the first char is a space.

ekohl commented on obal/data/modules/koji_find_tags.py:
```diff
@@ -0,0 +1,30 @@
+
+    for attributes in packages.values():
+        if 'koji_tags' in attributes:
+            for tag in attributes['koji_tags']:
+                tags.add(tag['name'])
```
What's the benefit of looping here instead of bulk adding a set?

ehelms replied:
I suppose if I put a generator in there to extract the `name` field I could do a bulk add?

ekohl replied:
Ah right, I missed it's extracting the name from `tag` and somehow assumed it was just adding `tag`. In that case, you can use this:
```python
tags += {tag['name'] for tag in attributes['koji_tags']}
```

Though I'm not sure which version states the intent more clearly.

@evgeni any opinion?

evgeni replied:
```
TypeError: unsupported operand type(s) for +=: 'set' and 'set'
```

:-)

`tags.update({…})` would work tho.

I'm fine with either form.

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,64 @@
+                    tag = koji_tag
+
+            if tag and 'package_base_dir' in attributes:
+                specfile = "%s/%s/%s.spec" % (attributes['package_base_dir'], package, package)
```
```suggestion
                specfile = os.path.join(attributes['package_base_dir'], package, "%s.spec".format(package))
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,64 @@
+
+            if tag and 'package_base_dir' in attributes:
+                specfile = "%s/%s/%s.spec" % (attributes['package_base_dir'], package, package)
+                scl = tag['scl'] if 'scl' in tag.keys() else None
```
```suggestion
                scl = tag.get('scl')
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,64 @@
+            executable=module.params['koji_executable']
+        )
+
+        packages_in_koji = set([item.split(' ')[0] for item in packages_in_koji.split("\n") if item])
```
You can use `set(generator)`:

```suggestion
        packages_in_koji = set(item.split(' ')[0] for item in packages_in_koji.split("\n") if item)
```

But shorter (and possibly faster) is set comprehension:

```suggestion
        packages_in_koji = {item.split(' ')[0] for item in packages_in_koji.split("\n") if item}
```

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,64 @@
+
+    try:
+        command = ['list-pkgs', '--quiet', '--tag', module.params['tag']]
+        packages_in_koji = koji(
```
I'd still rename this
```suggestion
        koji_output = koji(
```

ehelms replied:
I can. I just know its a common pattern to capture something and then down below is the same thing but just re-formatting it and using multiple variables to capture that seems overkill here.

ekohl replied:
I think it's a bad pattern. In Python 3 you have type hints which allows static code analysis using tools like [mypy](http://mypy-lang.org/). That's a good practice IMHO and helps you. It also allows IDEs to show more useful suggestions. Sadly, Ansible is incompatible with mypy for now, but it's good to learn to do it the right way:

```console
$ cat test.py 
import subprocess
def test() -> list:
    output = subprocess.check_output('x', universal_newlines=True)
    output = output.split()
    return output
$ mypy test.py 
test.py:6: error: Incompatible types in assignment (expression has type "List[str]", variable has type "str")
test.py:7: error: Incompatible return value type (got "str", expected "List[Any]")
Found 2 errors in 1 file (checked 1 source file)
```

Compare this to:
```console
$ cat test2.py 
import subprocess
def test() -> list:
    output = subprocess.check_output('x', universal_newlines=True)
    result = output.split()
    return result
$ mypy test2.py 
Success: no issues found in 1 source file
```

Now one cool thing it can do:
```console
 cat test3.py 
import subprocess
from typing import List
def test() -> List[str]:
    output = subprocess.check_output('x')
    result = output.split()
    return result
$ mypy test3.py 
test3.py:8: error: Incompatible return value type (got "List[bytes]", expected "List[str]")
Found 1 error in 1 file (checked 1 source file)
```

Note how it caught the missing `universal_newlines` parameter which causes encoding issues.

evgeni replied:
I tend to agree with ewoud here

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,64 @@
+    missing_in_git = packages_in_koji - packages_for_tag
+
+    module.exit_json(
+        changed=False,
```
Should this be true if there is a change?
```suggestion
        changed=missing_in_koji or missing_in_git,
```

Not sure this is semantically entirely correct though since it didn't make a change.

ekohl commented on obal/data/roles/verify_tag/tasks/verify.yaml:
```diff
@@ -0,0 +1,23 @@
+    msg: "{{ missing.missing_in_koji }}"
+  when: missing.missing_in_koji|length > 0
+
+- name: "Print packages missing in koji for {{ tag }}"
```
This doesn't look entirely appropriate, but I'm not sure what it should say.

ehelms replied:
In what way?

ekohl replied:
It doesn't actually print the packages missing, but rather tells you whether it was all ok. `Summarize koji vs sources status` or something feels closer, but it's still not that great.

ekohl commented:
> @ekohl Does this tag verification concept match what you want ? And is the command formatting OK?

Just wondering if it should be `verify-koji-tag`. It's also too bad we can't use positional arguments for the tags, but I guess that's just one limitation that we need to live with.

ehelms commented:
That is something I meant to ask @evgeni -- if there was a way in the structure of Obsah to have a command that has positional arguments that are not in the "inventory" ?

Another way around it I considered was having an inventory entry where tags were defined (under a "tags" group) and using that. I was pondering it related to the existence of [koji ansible modules](https://galaxy.ansible.com/ktdreyer/koji_ansible) and defining our tag structure rather than in tool_belt.

evgeni commented:
> That is something I meant to ask @evgeni -- if there was a way in the structure of Obsah to have a command that has positional arguments that are not in the "inventory" ?

Today? No.

But I think we could rather easily add a flag in the command metadata that says "the params for this playbook (command) aren't hosts (packages)". This would however mean that the playbook would get the params more or less verbatim and would have to process them itself.

ekohl commented:
The code that generates those is here:
https://github.com/theforeman/obsah/blob/439de5b10a16948d8fd3fbbff0840eaebd93a728/obsah/__init__.py#L300-L305

We can expose the nargs parameter here:
https://github.com/theforeman/obsah/blob/439de5b10a16948d8fd3fbbff0840eaebd93a728/obsah/__init__.py#L307-L309

That allows making tag a positional argument which is stored in an array. I don't really see a way to get the choices without writing Python, but that's probably less important.

ekohl commented:
> Another way around it I considered was having an inventory entry where tags were defined (under a "tags" group) and using that. I was pondering it related to the existence of [koji ansible modules](https://galaxy.ansible.com/ktdreyer/koji_ansible) and defining our tag structure rather than in tool_belt.

I think that could be nice. Might require some thought on our branching procedure + creating PRs in advance though.

ehelms commented:
The great irony, this change (https://github.com/theforeman/foreman-packaging/pull/5691) made turning tags into their own group of hosts much much harder as the hosts would need to have a variable in the name and that doesn't work well.

```
tags:
  hosts:
    "foreman-{{ foreman_version }}-rhel7":
      scl: 'tfm'
      dist: '.el7'
      macros:
        foremandist: "{{ foremandist }}"
    "foreman-{{ foreman_version }}-el8":
      dist: '.el8'
      macros:
        foremandist: "{{ foremandist }}"
```

Unless we made the tag entries more generic with configuration below, e.g.

```
tags:
  hosts:
    "foreman-rhel7":
      name: "foreman-{{ foreman_version }}-rhel7"
      scl: 'tfm'
      dist: '.el7'
      macros:
        foremandist: "{{ foremandist }}"
    "foreman-nonscl-rhel7":
      name: "foreman-{{ foreman_version }}-nonscl-rhel7"
      scl: 'tfm'
      dist: '.el7'
      macros:
        foremandist: "{{ foremandist }}"
    "foreman-el8":
      name: "foreman-{{ foreman_version }}-el8"
      dist: '.el8'
      macros:
        foremandist: "{{ foremandist }}"
```

Further downside to hosts entry method, I'm ot sure how to then translate that for easy inclusion as a re-usable unit into the vars for package entries. For example, we define the set of tags associated to a set of packages as a variable. 

```
    koji_tags:
      - name: "foreman-{{ foreman_version }}-rhel7"
        scl: 'tfm'
        dist: '.el7'
        macros:
          foremandist: "{{ foremandist }}"
      - name: "foreman-{{ foreman_version }}-el8"
        dist: '.el8'
        macros:
          foremandist: "{{ foremandist }}"
```

evgeni commented:
@ehelms so to summarize the discussion above (because I feel a bit lost):

The code in this PR works and the current invocation looks like this:
```console
obal verify-tag --tag foreman-nightly-rhel7
obal verify-tag --all
```

You'd prefer to have this be:
```console
obal verify-tag foreman-nightly-rhel7
obal verify-tag (--)?all
```
but that's not possible due to 1) the way you redefined the version numbering in the manifest 2) us not having a nice way to create a command that takes a positional argument that is not a host?

I can take a stab at 2) but I don't think this should block this PR as the current invocation is "good enough"

evgeni commented:
Okay, that was easier than expected: https://github.com/theforeman/obsah/pull/6

And no python involved to pull out the data as it gets passed to ansible via `-e blah=blah`

evgeni commented on obal/data/playbooks/verify-tag/metadata.obal.yaml:
```diff
@@ -0,0 +1,8 @@
+help: Verify packages against a tag
+variables:
+  tag:
+    help: Tag to verify against
```
with https://github.com/theforeman/obsah/pull/6 merged, you can now do:

```suggestion
  tag:
    parameter: tag
    help: Tag to verify against
```

and you'll get a positional argument

mind you, positional arguments are required, so you'll probably want to drop the `verify_all` arg and specialcase on an `all` tag or something

ehelms replied:
Can you release a new Obsah to make testing easier? Also, what does ` specialcase on an all tag or something` mean?

evgeni replied:
https://github.com/theforeman/obal/pull/243/files#r486301398 + https://github.com/theforeman/obal/pull/243/files#r486301876

evgeni commented on obal/data/playbooks/verify-tag/metadata.obal.yaml:
```diff
@@ -0,0 +1,8 @@
+variables:
+  tag:
+    help: Tag to verify against
+  verify_tag_all:
+    action: store_true
+    parameter: --all
+    help: Verify all tags
```
```suggestion
```

evgeni commented on obal/data/roles/verify_tag/tasks/main.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+- include_tasks: "verify.yaml"
+  when: not verify_tag_all
+
+- include_tasks: "all.yaml"
+  when: verify_tag_all
```
```suggestion
- include_tasks: "verify.yaml"
  when: tag != 'all'

- include_tasks: "all.yaml"
  when: tag == 'all'
```

ehelms commented:
Testing locally this seemed to work with the positional args, thanks for tackling that @evgeni ! Now to get these lint failures sorted

evgeni commented:
I've prepared a new obsah release in https://github.com/theforeman/obsah/pull/7 -- can you please bump the requirement of it in setup.py to >=0.0.2 when it's released?

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,64 @@
+            executable=module.params['koji_executable']
+        )
+
+        packages_in_koji = {item.split(' ')[0] for item in packages_in_koji.split("\n") if item}
```
Nit: I always like to limit the number of splits if I can. Feels more efficient to me though I never really measured it.
```suggestion
        packages_in_koji = {item.split(' ', 1)[0] for item in packages_in_koji.split("\n") if item}
```

evgeni commented on tests/fixtures/help/verify-tag.txt:
```diff
@@ -0,0 +1,14 @@
+usage: obal verify-tag [-h] [-v] [-e EXTRA_VARS] [--tag TAG] [--all]
```
this file needs updates now

evgeni commented on tests/test_functional.py:
```diff
@@ -670,3 +674,22 @@ def test_obal_mock():
     assert_mockbin_log(expected_log)
+
+
+@obal_cli_test(repotype='upstream')
+def test_obal_verify_tag():
+    assert_obal_success(['verify-tag', '--tag', 'obaltest-nightly-rhel7'])
+
+    expected_log = ['koji list-pkgs --quiet --tag obaltest-nightly-rhel7']
+    assert_mockbin_log(expected_log)
+
+
+@obal_cli_test(repotype='upstream')
+def test_obal_verify_tag_all():
+    assert_obal_success(['verify-tag', '--all'])
+
+    expected_log = [
+        'koji list-pkgs --quiet --tag obaltest-nightly-el8',
+        'koji list-pkgs --quiet --tag obaltest-nightly-rhel7'
+    ]
+    assert_mockbin_log(expected_log)
```
these tests need updates now

evgeni commented on tests/test_functional.py:
```diff
@@ -670,3 +674,22 @@ def test_obal_mock():
+
+@obal_cli_test(repotype='upstream')
+def test_obal_verify_tag():
+    assert_obal_success(['verify-tag', '--tag', 'obaltest-nightly-rhel7'])
```
```suggestion
    assert_obal_success(['verify-tag', 'obaltest-nightly-rhel7'])
```

evgeni commented on tests/test_functional.py:
```diff
@@ -670,3 +674,22 @@ def test_obal_mock():
+
+@obal_cli_test(repotype='upstream')
+def test_obal_verify_tag_all():
+    assert_obal_success(['verify-tag', '--all'])
```
```suggestion
    assert_obal_success(['verify-tag', 'all'])
```

ekohl reviewed (changes requested):
This also needs to set the required obsah version to at least 0.0.2

ehelms commented:
Updated addressing all comments, this needs https://github.com/theforeman/obal/pull/290

ehelms commented:
Rebased and updated

ehelms commented:
Current output as of this comment:

```
obal-source verify-koji-tag all

PLAY [localhost] *************************************************************************************************************************************************************************************************************************************************************************
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/all.yaml for localhost

TASK [verify_koji_tag : Find all tags] ***************************************************************************************************************************************************************************************************************************************************
ok: [localhost]
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost

TASK [verify_koji_tag : Verify tag] ******************************************************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [verify_koji_tag : Packages in Koji but not in git for foreman-client-nightly-el8] **************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg:
  - jsoncpp
  - qpid-proton

TASK [verify_koji_tag : Verify tag] ******************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [verify_koji_tag : Print packages missing in koji for foreman-client-nightly-el9] ***************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg: Tag foreman-client-nightly-el9 verified

TASK [verify_koji_tag : Verify tag] ******************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [verify_koji_tag : Print packages missing in koji for foreman-client-nightly-rhel7] *************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg: Tag foreman-client-nightly-rhel7 verified

TASK [verify_koji_tag : Verify tag] ******************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [verify_koji_tag : Print packages missing in koji for foreman-nightly-el8] **********************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg: Tag foreman-nightly-el8 verified

TASK [verify_koji_tag : Verify tag] ******************************************************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [verify_koji_tag : Packages in Koji but not in git for foreman-plugins-nightly-el8] *************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg:
  - yggdrasil

TASK [verify_koji_tag : Verify tag] ******************************************************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [verify_koji_tag : Packages in Koji but not in git for katello-nightly-el8] *********************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg:
  - rubygem-pulp_2to3_migration_client
  - rubygem-redhat_access
  - rubygem-redhat_access_lib

TASK [verify_koji_tag : Packages missing in koji for katello-nightly-el8] ****************************************************************************************************************************************************************************************************************
ok: [localhost] => 
  msg:
  - rubygem-qpid_proton
```

ekohl commented on obal/data/modules/koji_find_tags.py:
```diff
@@ -0,0 +1,29 @@
+
+def main():
+    """
+    Verify a tag
```
This doesn't match the rest

ekohl commented on obal/data/modules/koji_verify_tag.py:
```diff
@@ -0,0 +1,64 @@
+    missing_in_koji = packages_for_tag - packages_in_koji
+    missing_in_git = packages_in_koji - packages_for_tag
+
+    module.exit_json(
```
Can this fail `if missing_in_koji or missing_in_git`? Ansible has a way to continue on error, right? Or perhaps the regular failure output is sufficient anyway.

ehelms replied:
Yes there are two options:

 1) Have the module fail and allow users of the module to ignore errors if they want
 2) Have this module simply report data and then have a task that fails conditionally

I do not know what is more correct, so I am Ok with either strategically.

ekohl replied:
I'd lean to option 1.

ekohl reviewed (commented):
Overall this looks good.

ehelms commented:
New output:

```
obal-source verify-koji-tag all

PLAY [localhost] *************************************************************************************************************************************************************************************************************************************************************************
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/all.yaml for localhost

TASK [verify_koji_tag : Find all tags] ***************************************************************************************************************************************************************************************************************************************************
ok: [localhost]
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost
included: /home/ehelms/workspace/upstream/packaging/obal/obal/data/roles/verify_koji_tag/tasks/verify.yaml for localhost

TASK [verify_koji_tag : Verify tag] ******************************************************************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => changed=true 
  missing_in_git:
  - jsoncpp
  - qpid-proton
  missing_in_koji: []
  msg: Package differences found

PLAY RECAP *******************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=8    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   
```

ekohl reviewed (commented):
User wise :+1: but CI is unhappy.

ehelms commented:
@evgeni You did some initial reviews, do you care to have another look? Otherwise I will merge this.
